### ISSUE TITLE
NoMethodError: undefined method `to_sym' for #<PDF::Reader::Stream:0x007fe081ca3d70>

### ISSUE THREAD

**Issue by MathieuDerelle** [yob/pdf-reader#279]
this PDF raises an error when trying to get its text content

[20180809-1650-3297-517320-1.pdf](https://github.com/yob/pdf-reader/files/2278367/20180809-1650-3297-517320-1.pdf)

```
[7] pry(main)> r=PDF::Reader.new(path);
[8] pry(main)> r.pages.first.text
NoMethodError: undefined method `to_sym' for #<PDF::Reader::Stream:0x007fe080298e80>
```

backtrace : 

```
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/encoding.rb:47:in `initialize'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/font.rb:112:in `new'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/font.rb:112:in `extract_base_info'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/font.rb:44:in `initialize'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:384:in `new'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:384:in `block in build_fonts'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:383:in `each'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:383:in `map'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:383:in `build_fonts'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_state.rb:29:in `initialize'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_text_receiver.rb:43:in `new'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page_text_receiver.rb:43:in `page='
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page.rb:161:in `block in callback'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page.rb:160:in `each'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page.rb:160:in `callback'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page.rb:111:in `walk'
/Users/mathieu/.rvm/gems/ruby-2.2.5/bundler/gems/pdf-reader-7fa71990d83c/lib/pdf/reader/page.rb:82:in `text'
(pry):10:in `__pry__'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:355:in `eval'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:355:in `evaluate_ruby'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:323:in `handle_line'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:243:in `block (2 levels) in eval'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:242:in `catch'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:242:in `block in eval'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:241:in `catch'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_instance.rb:241:in `eval'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:77:in `block in repl'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:67:in `loop'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:67:in `repl'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:38:in `block in start'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/input_lock.rb:61:in `call'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/input_lock.rb:61:in `__with_ownership'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/input_lock.rb:79:in `with_ownership'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:38:in `start'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/repl.rb:15:in `start'
/Users/mathieu/.rvm/gems/ruby-2.2.5/gems/pry-0.10.4/lib/pry/pry_class.rb:169:in `start'
bin/console:13:in `<main>'
```

as you can see in the backtrace, it is happening even on master

**Comment by redrick** [comment]
same here, on stable `2.1.0`:

my difference is that I am opening StringIO from railses own active storage... no idea if that helps, seems more like a problem of `page()` method, but what is weird is that I ran it through sidekiq on background and failed twice and on third time went through ok.... so I only got this from sentry afterwards...

Any ideas, what to try/provide to find fixable problem here?

```ruby
pdf = PDF::Reader.new(StringIO.new(message.files.first.download))
pdf.page(1)
```
```ruby
NoMethodError: undefined method `to_sym' for #<PDF::Reader::Stream:0x007efd0c0ba508>
Did you mean?  to_s
  from pdf/reader/encoding.rb:46:in `initialize'
  from pdf/reader/font.rb:111:in `new'
  from pdf/reader/font.rb:111:in `extract_base_info'
  from pdf/reader/font.rb:43:in `initialize'
  from pdf/reader/page_state.rb:383:in `new'
  from pdf/reader/page_state.rb:383:in `block in build_fonts'
  from pdf/reader/page_state.rb:382:in `each'
  from pdf/reader/page_state.rb:382:in `map'
  from pdf/reader/page_state.rb:382:in `build_fonts'
  from pdf/reader/page_state.rb:28:in `initialize'
  from pdf/reader/page_text_receiver.rb:42:in `new'
  from pdf/reader/page_text_receiver.rb:42:in `page='
  from pdf/reader/page.rb:160:in `block in callback'
  from pdf/reader/page.rb:159:in `each'
  from pdf/reader/page.rb:159:in `callback'
  from pdf/reader/page.rb:110:in `walk'
  from pdf/reader/page.rb:81:in `text'
  from app/services/isds/data_messages/assigner.rb:37:in `try_find_vat'
  from app/services/isds/data_messages/assigner.rb:21:in `assign'
  from app/services/isds/data_messages/assigner.rb:14:in `assign'
  from app/services/isds/data_boxes/downloader.rb:79:in `assign_account_for'
  from app/services/isds/data_boxes/downloader.rb:65:in `block (3 levels) in <class:Downloader>'
  from app/services/isds/data_boxes/downloader.rb:62:in `each'
  from app/services/isds/data_boxes/downloader.rb:62:in `block (2 levels) in <class:Downloader>'
  from app/services/isds/data_boxes/downloader.rb:30:in `download_received'
  from app/services/isds/data_boxes/downloader.rb:23:in `download'
  from app/services/isds/data_boxes/downloader.rb:18:in `download'
  from app/services/isds/data_boxes/batch_downloader.rb:30:in `download'
  from app/services/isds/data_boxes/batch_downloader.rb:19:in `block in download_all'
  from app/services/isds/data_boxes/batch_downloader.rb:18:in `each'
  from app/services/isds/data_boxes/batch_downloader.rb:18:in `download_all'
  from app/services/isds/data_boxes/batch_downloader.rb:14:in `download_all'
  from app/jobs/download_all_databoxes_job.rb:5:in `perform'
  from active_job/execution.rb:39:in `block in perform_now'
  from active_support/callbacks.rb:109:in `block in run_callbacks'
  from raven/integrations/rails/active_job.rb:18:in `capture_and_reraise_with_sentry'
  from raven/integrations/rails/active_job.rb:12:in `block (2 levels) in included'
  from active_support/callbacks.rb:118:in `instance_exec'
  from active_support/callbacks.rb:118:in `block in run_callbacks'
  from i18n.rb:268:in `with_locale'
  from active_job/translation.rb:9:in `block (2 levels) in <module:Translation>'
  from active_support/callbacks.rb:118:in `instance_exec'
  from active_support/callbacks.rb:118:in `block in run_callbacks'
  from active_job/logging.rb:26:in `block (4 levels) in <module:Logging>'
  from active_support/notifications.rb:168:in `block in instrument'
  from active_support/notifications/instrumenter.rb:23:in `instrument'
  from active_support/notifications.rb:168:in `instrument'
  from active_job/logging.rb:25:in `block (3 levels) in <module:Logging>'
  from active_job/logging.rb:46:in `block in tag_logger'
  from active_support/tagged_logging.rb:71:in `block in tagged'
  from active_support/tagged_logging.rb:28:in `tagged'
  from active_support/tagged_logging.rb:71:in `tagged'
  from active_job/logging.rb:46:in `tag_logger'
  from active_job/logging.rb:22:in `block (2 levels) in <module:Logging>'
  from active_support/callbacks.rb:118:in `instance_exec'
  from active_support/callbacks.rb:118:in `block in run_callbacks'
  from active_support/callbacks.rb:136:in `run_callbacks'
  from active_job/execution.rb:35:in `perform_now'
  from active_job/execution.rb:24:in `block in execute'
  from active_support/callbacks.rb:109:in `block in run_callbacks'
  from active_job/railtie.rb:28:in `block (4 levels) in <class:Railtie>'
  from active_support/execution_wrapper.rb:87:in `wrap'
  from active_support/reloader.rb:73:in `block in wrap'
  from active_support/execution_wrapper.rb:83:in `wrap'
  from active_support/reloader.rb:72:in `wrap'
  from active_job/railtie.rb:27:in `block (3 levels) in <class:Railtie>'
  from active_support/callbacks.rb:118:in `instance_exec'
  from active_support/callbacks.rb:118:in `block in run_callbacks'
  from active_support/callbacks.rb:136:in `run_callbacks'
  from active_job/execution.rb:22:in `execute'
  from active_job/queue_adapters/sidekiq_adapter.rb:42:in `perform'
  from sidekiq/processor.rb:167:in `execute_job'
  from sidekiq/processor.rb:139:in `block (5 levels) in process'
  from sidekiq/rails.rb:90:in `block in call'
  from active_support/execution_wrapper.rb:87:in `wrap'
  from sidekiq/rails.rb:89:in `call'
  from sidekiq/processor.rb:135:in `block (4 levels) in process'
  from sidekiq/middleware/chain.rb:128:in `block in invoke'
  from lib/sidekiq/middleware/server/sentry_error_logger.rb:9:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from raven/integrations/sidekiq.rb:9:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/server/logging.rb:10:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/server/retry_jobs.rb:74:in `call'
  from sidekiq/middleware/chain.rb:130:in `block in invoke'
  from sidekiq/middleware/chain.rb:133:in `invoke'
  from sidekiq/processor.rb:134:in `block (3 levels) in process'
  from sidekiq/logging.rb:32:in `with_context'
  from sidekiq/processor.rb:132:in `block (2 levels) in process'
  from sidekiq/processor.rb:183:in `stats'
  from sidekiq/processor.rb:131:in `block in process'
  from sidekiq.rb:35:in `block in <module:Sidekiq>'
  from sidekiq/processor.rb:126:in `process'
  from sidekiq/processor.rb:82:in `process_one'
  from sidekiq/processor.rb:70:in `run'
  from sidekiq/util.rb:17:in `watchdog'
  from sidekiq/util.rb:26:in `block in safe_thread'
```

**Comment by MathieuDerelle** [comment]
I'm not quite sure why but this issue is solved and this gem as not been updated at all

**Comment by MathieuDerelle** [comment]
Depending on the project in which I use this gem, it works or not.

I have one project where PDF::Reader can extract the text without issue and the other where it can't

The problem is that I could not figure out why : 
- I've copied the Gemfile, did not change anything.
- Made sure they both run on same ruby version, did not change anything.

In the working project, there is still a strange line extracted at the beginning of each page : 
`%META=C00245252;09091539;N;20180228;0;1180228;09;N;O;2018/09F1802004;C00245252_09091539.pdf%`

**Comment by pierre-alain-b** [comment]
I have the same issue here with 2.2.1 and 2.3.0.

**Comment by pierre-alain-b** [comment]
If I "clean" the PDF with this command beforehand, it can be parsed with no issue.
gs -o output.pdf -sDEVICE=pdfwrite input.pdf

**Comment by yob** [comment]
Thanks for the bug report, and sorry it took me a while to get to it.

I've addressed the crash in #313, which I'll release in a gem soon.

I'm not 100% confident that I'm interpreting the characters in CMap based fonts 100% correctly, but it's correct for the sample file provided by @MathieuDerelle. Hopefully that's good enough to start with, and I can fix any accuracy issues that come up if/when they're raised as issues.

**Comment by pierre-alain-b** [comment]
Thank you, much appreciated! It seems to work on my side as well.
